From e25af4d0a6bb534a1bb9636774e810033f6c30ae Mon Sep 17 00:00:00 2001
From: Ahmad Shtewe <ahmad.shtewe@solid-run.com>
Date: Sat, 1 Mar 2025 12:49:01 +0100
Subject: [PATCH 14/15] lx2160a: fix i2c bus flushing glitch

There is an unintended glitch when setting gpio direction register
before data register.

Reorder the gpio initialisation to set direction last during setup, and
first during cleanup before restoring original values.

Signed-off-by: Ahmad Shtewe <ahmad.shtewe@solid-run.com>
Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 plat/nxp/common/setup/ls_i2c_init.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/plat/nxp/common/setup/ls_i2c_init.c b/plat/nxp/common/setup/ls_i2c_init.c
index 76dfd3376..591ede2f7 100644
--- a/plat/nxp/common/setup/ls_i2c_init.c
+++ b/plat/nxp/common/setup/ls_i2c_init.c
@@ -221,9 +221,12 @@ static void ls_i2c_flush(uint8_t busno, const char *busname) {
 	gpdir |= scl_mask | sda_mask;
 	gpodr |= scl_mask | sda_mask;
 	gpdat |= scl_mask | sda_mask;
-	mmio_write_32(gpdir_addr, gpdir);
-	mmio_write_32(gpodr_addr, gpodr);
 	mmio_write_32(gpdat_addr, gpdat);
+	mmio_write_32(gpodr_addr, gpodr);
+	mmio_write_32(gpdir_addr, gpdir);
+
+	/* allow short delay for changes to propagate */
+	udelay(10);
 
 	/*
 	 * reliable detection of blocked bus is hard
@@ -242,9 +245,9 @@ static void ls_i2c_flush(uint8_t busno, const char *busname) {
 	}
 
 	/* restore configuration registers */
-	mmio_write_32(gpdat_addr, backup.gpdat);
-	mmio_write_32(gpodr_addr, backup.gpodr);
 	mmio_write_32(gpdir_addr, backup.gpdir);
+	mmio_write_32(gpodr_addr, backup.gpodr);
+	mmio_write_32(gpdat_addr, backup.gpdat);
 	mmio_write_32(info->pinmux_addr, backup.pinmux);
 
 	return;
-- 
2.43.0

